Add pre-commit hook and CI check, excluding currently unformatted files (#7745)
NixOS/nix: doc/manual/src/contributing/hacking.md

$ nix develop
```

## Pre-commit and formatting

> `nix develop` installs Nix-[integrated](https://github.com/cachix/pre-commit-hooks.nix) [pre-commit](https://pre-commit.com) hooks.

Reviewer: This should be turned off by default. git hooks can be incredible annoying when they are not fully working and you are in the middle of a more complicated git action like a rebase or you are doing a commit which is broken on purpose and will be later fixed. Also when switching to an old branch when still inside the nix develop shell, this will probably blow up.

Reviewer 2: I think it's OK. It's easy enough to use `git commit -n` to skip the Git hooks when needed.

It could be nice to have an environment variable to configure this behavior, like `NIX_DEVELOP_PRE_COMMIT`.

The formatters will run when you commit.
Pay attention to the console output. If it fails, run `git add --patch` to
approve the suggestions _and commit again_.

> The hook configuration is only regenerated when you start a new dev shell.

Reviewer 3: ```suggestion
The hook configuration is only regenerated when you start a new development shell.
```

> Some other editors (e.g. Emacs, Vim) need a plugin to support LSP servers in general (e.g. [lsp-mode](https://github.com/emacs-lsp/lsp-mode) for Emacs and [vim-lsp](https://github.com/prabirshrestha/vim-lsp) for vim).
> Editor-specific setup is typically opinionated, so we will not cover it here in more detail.

> ## Formatting and pre-commit

Reviewer 3: ```suggestion
## Formatting and pre-commit hooks
```

You may run the formatters as a one-off using:

```console
make format
```

If you'd like to run the formatters before every commit, install the hooks:

```
pre-commit-hooks-install
```

This installs [pre-commit](https://pre-commit.com) using [cachix/git-hooks.nix](https://github.com/cachix/git-hooks.nix).

When making a commit, pay attention to the console output.
If it fails, run `git add --patch` to approve the suggestions _and commit again_.

> To refresh the config, do the following:

Reviewer 3: Which config? The dev shell?

> - if you use `make format`: stop and start your `nix develop` shell.

Reviewer 3: ```suggestion
- if you use `make format`: exit the development shell and start it again by running `nix develop`.
```

> - if you use the pre-commit hook: stop and start, and run `pre-commit-hooks-install` again.

Reviewer 3: Stop and start what?
